examples/advanced_operations/add_ad_customizer.rb

```diff
@@ -0,0 +1,383 @@
+  )
+
+  # Set the feed to be used only with the specified ad groups.
+  create_feed_item_targets(
+    client,
+    customer_id,
+    ad_group_ids,
+    feed_item_resource_names,
+  )
+
+  # Create ads that use the feed for customization.
+  create_ads_with_customizations(
+    client,
+    customer_id,
+    ad_group_ids,
+    feed_name,
+  )
+end
+
+# Creates a feed to be used for ad customization.
+def create_ad_customizer_feed(
+  client, customer_id, feed_name)
+  # Creates a feed operation for creating a feed.
+  operation = client.operation.create_resource.feed do |feed|
+    feed.name = feed_name
+    # Creates three feed attributes: a name, a price and a date. The attribute
+    # names are arbitrary choices and will be used as placeholders in the ad
+    # text fields.
+    feed.attributes << client.resource.feed_attribute do |a|
+      a.type = :STRING
+      a.name = "Name"
+    end
+    feed.attributes << client.resource.feed_attribute do |a|
+      a.type = :STRING
+      a.name = "Price"
+    end
+    feed.attributes << client.resource.feed_attribute do |a|
+      a.type = :DATE_TIME
+      a.name = "Date"
+    end
+    feed.origin = :USER
+  end
+
+  # Issues a mutate request to add the feed.
+  feed_response = client.service.feed.mutate_feeds(
+    customer_id: customer_id,
+    operations: [operation],
+  )
+
+  feed_resource_name = feed_response.results.first.resource_name
+  puts "Added a feed with resource name #{feed_resource_name}."
+
+  feed_resource_name
+end
+
+# Retrieves attributes for a feed.
+def get_feed_attributes(
+  client, customer_id, ad_customizer_feed_resource_name)
+  query = <<~QUERY
+    SELECT feed.attributes, feed.name
+    FROM feed
+    WHERE feed.resource_name = "#{ad_customizer_feed_resource_name}"
+  QUERY
+
+  response = client.service.google_ads.search(
+    customer_id: customer_id,
+    query: query,
+    page_size: PAGE_SIZE,
+  )
+
+  feed = response.first.feed
+  feed_details = {}
+  puts "Found the following attributes for feed with name #{feed.name}:"
+  feed.attributes.each do |a|
+    feed_details[a.name.value] = a.id
+    puts "\t#{a.name} with id #{a.id} and type #{a.type}"
+  end
+
+  feed_details
+end
+
+# Creates a feed mapping for a given feed.
+def create_ad_customizer_mapping(
+  client,
+  customer_id,
+  ad_customizer_feed_resource_name,
+  ad_customizer_feed_attributes)
+  # Creates the operation.
+  operation = client.operation.create_resource.feed_mapping do |fm|
+    fm.placeholder_type = :AD_CUSTOMIZER
+    fm.feed = ad_customizer_feed_resource_name
+    # Maps the feed attribute IDs to the field ID constants.
+    fm.attribute_field_mappings << client.resource.attribute_field_mapping do |af|
+      af.feed_attribute_id = ad_customizer_feed_attributes["Name"]
+      af.ad_customizer_field = :STRING
+    end
+    fm.attribute_field_mappings << client.resource.attribute_field_mapping do |af|
+      af.feed_attribute_id = ad_customizer_feed_attributes["Price"]
```

Comment:
You should use symbols instead of strings as hash keys whenever you can. Strings in Ruby are mutable, so every time you hard code a string it uses new memory. Symbols are immutable, so multiple of the same symbol use the same memory. You can use "to_sym" to convert a string to a symbol.

Reply:
Switched to using symbols.